travis CI: add test for cmake build

WeinaJi opened the pull request.

nrnhines committed:
Array for spike compression may not be allocated in some circumstances. (#293)

If normal uncompressed spike exchange occurs, then a later request
for compressed spike exchange will fail to allocate an overflow array
and access of a NULL pointer will occur

pramodk committed:
Update coreneuron submodule and patch hh.mod if coreneuron is enabled

adamjhn committed:
bugfix: exception when deleting a species/parameter before initialization. (#295)

pramodk committed:
Remove legacy options for purify and mpitrace library of bg-q (#296)

pramodk committed:
Avoid bash warning/error for empty if block (#298)

* Avoid bash warning/error for empty if block

* replace : with false

Commit:
travis CI: add test for cmake build

orena1 committed:
update copyright (#300)

WeinaJi commented on src/nrniv/nrnpy.cpp:
```diff
@@ -372,21 +371,31 @@ static void* load_nrnpython_helper(const char* npylib) {
-#if __MINGW32__
-static int digittoint(char c) {
-	return int(c - '0');
+int digit_to_int(char ch) {
```
The function "diggittoint" doesn't exist in ctype.h in linux, so here I copy the definition from BSD and rename it. pylib2pyver10 will call this new function.

pramodk committed:
Merge branch 'cmake' into travis

pramodk committed:
explicitly turn on/off interviews

pramodk committed:
typo and minor fixes : can't have comment in if-else list item

pramodk committed:
update submodules : iv as well as coreneuron (#302)

pramodk commented on .travis.yml:
```diff
@@ -10,35 +10,76 @@ addons:
+      cmake ${CMAKE_OPTION} -DCMAKE_INSTALL_PREFIX=$(pwd)/install ..;
+      cmake --build .;
+      make install;
+      echo -e 'topology()' | "$(pwd)/install/bin/neurondemo" | grep -q soma;
```
here we need to run `$PYTHON -c 'import neuron; neuron.test()';` if python is enabled. Also, in case of multiple pythons are enabled, we should test multiple pythons.

lets discuss offline.

Commit:
travis CI: add python tests for cmake build

pramodk committed:
CMake cleanup : remove unnecessary log/output messages

Commit:
travis CI

nrnhines committed:
Merge branch 'master' into cmake

nrnhines committed:
add back the run_all.py travis_cl test

nrnhines committed:
Merge pull request #304 from neuronsimulator/cmake-1

Cmake 1

pramodk commented on .travis.yml:
```diff
@@ -10,35 +10,91 @@ addons:
+      export PYTHONPATH=$PYTHONPATH:$(pwd)/build/install/lib/python/;
+    else
+      export PYTHONPATH=$PYTHONPATH:$(pwd)/install/lib/python/;
+    fi
```
Can't we just choose same install path? e.g. $HOME/install ?

WeinaJi replied:
Done with INSTALL_DIR="$HOME/install"

pramodk commented on .travis.yml:
```diff
@@ -10,35 +10,91 @@ addons:
+    else
+      export PYTHONPATH=$PYTHONPATH:$(pwd)/install/lib/python/;
+    fi
+  - if [ "$NRN_ENABLE_PYTHON_DYNAMIC" == "ON" ]; then
+      pyenv global 2.7.15 3.7.1;
+    else
+      pyenv global $PYTHON_VERSION;
+    fi
```
nitpick : what if we alway enable both python environments ? i.e. `pyenv global 2.7.15 3.7.1;`

WeinaJi replied:
In principle we can enable both and cmake will always take the first version in pyenv global to build. I would prefer to keep the single version for normal build here so that we know clearly what is to be used.

pramodk replied:
sounds reasonable.

pramodk commented on .travis.yml:
```diff
@@ -10,35 +10,91 @@ addons:
+      $PIP install --user cython;
+    else
+      $PIP install --user scipy matplotlib bokeh ipython cython --upgrade;
+    fi
```
nitpick : if these packages take minimal time to install, can we install them all ? (just thinking of simplifying logic in the script)

WeinaJi replied:
Now we install them all for both cases.

pramodk commented on .travis.yml:
```diff
@@ -10,35 +10,91 @@ addons:
+        CMAKE_OPTION="-DNRN_ENABLE_PYTHON=ON -DNRN_ENABLE_PYTHON_DYNAMIC=ON -DNRN_PYTHON_DYNAMIC=${PYTHON2};${PYTHON3}";
+      fi;
+      mkdir build && cd build;
+      cmake ${CMAKE_OPTION} -DCMAKE_INSTALL_PREFIX=$(pwd)/install ..;
```
`$(pwd)/install` can be `$HOME/install`?

WeinaJi replied:
Done

pramodk commented on .travis.yml:
```diff
@@ -10,35 +10,91 @@ addons:
+      make install;
+    else
+      export PATH=$(pwd)/install/x86_64/bin:$PATH;
+      ./travis_build.sh .;
```
and you can also update travis_build.sh to install in `$HOME/install`

WeinaJi replied:
Done

pramodk commented on .travis.yml:
```diff
@@ -10,35 +10,91 @@ addons:
+    fi
+  - if [ "$BUILD_MODE" == "cmake" ]; then
+      $(pwd)/install/bin/neurondemo -nogui -c 'demo(4)' -c 'run()' -c 'quit()';
+    fi
```
this should also work in autoconf build. Isn't it? (may be path in autoconf os install/x86_64/bin/neurondemo)

WeinaJi replied:
Yes, add it for autoconf as well. path is not mandatory because it's has been added into the $PATH .

pramodk commented on .travis.yml:
```diff
@@ -10,35 +10,91 @@ addons:
+      $(pwd)/install/bin/neurondemo -nogui -c 'demo(4)' -c 'run()' -c 'quit()';
+    fi
+  - if [ "$NRN_ENABLE_PYTHON_DYNAMIC" == "ON" ]; then
+      $PYTHON2 --version;
```
nitpick : can be removed?

WeinaJi replied:
Done

pramodk commented on .travis.yml:
```diff
@@ -10,35 +10,91 @@ addons:
+  - if [ "$NRN_ENABLE_PYTHON_DYNAMIC" == "ON" ]; then
+      $PYTHON2 --version;
+      $PYTHON2 -c 'import neuron; neuron.test()';
+      $PYTHON3 --version;
```
nitpick : can be removed?

pramodk commented on .travis.yml:
```diff
@@ -10,35 +10,91 @@ addons:
+      $PYTHON3 -c 'import neuron; neuron.test()';
+    else
+      if [ "$CMAKE_OPTION" != "-DNRN_ENABLE_PYTHON=OFF" ]; then
+        $PYTHON --version;
```
same

WeinaJi replied:
Done

pramodk reviewed (commented):
excellent!

I have just few minor comments.

Commit:
travis CI : reformat

Commit:
travis CI: fix the python version

pramodk committed:
Minor cosmetic changes

pramodk committed:
Merge branch 'yale_cmake' into travis

pramodk committed:
move tests inside cmake section

pramodk committed:
Revert "move tests inside cmake section"

This reverts commit 29c40566e27485848b3c339fe76d594c2b8e447c.

pramodk committed:
Fix rxd tests with CMake
  - use sys.executable to get path of python used to launch tests (this
  is because on my machine there was no python2 but python2.7. So using
  sys.executable seems more robust?)
  - build librxdmath
  - run rxd tests on travis

Commit:
travis CI: revert changes

Commit:
travis CI for cmake build

pramodk commented:
@WeinaJi  : I think the command status is not returned correct because of multiple statements (using `;`). For example [see this build](https://travis-ci.org/BlueBrain/nrn/jobs/609778015?utm_medium=notification&utm_source=github_status) which has:

```
9 warnings generated.
2154/usr/local/clang-7.0.0/bin/clang++ -pthread -shared -L/opt/python/2.7.15/lib -Wl,-rpath=/opt/python/2.7.15/lib build/temp.linux-x86_64-2.7/inithoc.o -L/home/travis/build/BlueBrain/nrn/build/lib -L/home/travis/build/BlueBrain/nrn/external/iv/src/include/../lib -L/opt/python/2.7.15/lib -lnrnpython2 -lnrniv -linterviews -lpython2.7 -o build/lib.linux-x86_64-2.7/neuron/hoc.so -Wl,-R/home/travis/build/BlueBrain/nrn/build/lib -Wl,-R/home/travis/build/BlueBrain/nrn/external/iv/src/include/../lib
2155/usr/bin/ld: cannot find -lnrnpython2
2156/usr/bin/ld: cannot find -linterviews
2157clang-7: error: linker command failed with exit code 1 (use -v to see invocation)
2158error: command '/usr/local/clang-7.0.0/bin/clang++' failed with exit status 1
2159src/nrnpython/CMakeFiles/hoc_module.dir/build.make:71: recipe for target 'hoc_module' failed
2160make[2]: *** [hoc_module] Error 1
2161CMakeFiles/Makefile2:1917: recipe for target 'src/nrnpython/CMakeFiles/hoc_module.dir/all' failed
2162make[1]: *** [src/nrnpython/CMakeFiles/hoc_module.dir/all] Error 2
2163Makefile:129: recipe for target 'all' failed
2164make: *** [all] Error 2
2165The command "if [ "$BUILD_MODE" == "cmake" ]; then if [ "$NRN_ENABLE_PYTHON_DYNAMIC" == "ON" ]; then export PYTHON2=$(pyenv which python2.7); export PYTHON3=$(pyenv which python3.7); CMAKE_OPTION="-DNRN_ENABLE_PYTHON=ON -DNRN_ENABLE_PYTHON_DYNAMIC=ON -DNRN_PYTHON_DYNAMIC=${PYTHON2};${PYTHON3}"; fi; mkdir build && cd build; cmake $CMAKE_OPTION -DCMAKE_INSTALL_PREFIX=$INSTALL_DIR ..; cmake --build . -- -j 2; make install; cd ..; export PATH=$INSTALL_DIR/bin:$PATH; else export PATH=$INSTALL_DIR/x86_64/bin:$PATH; ./travis_build.sh $INSTALL_DIR; fi" exited with 0.
21660.21s$ neurondemo -nogui -c 'demo(4)' -c 'run()' -c 'quit()'; if [ "$NRN_ENABLE_PYTHON_DYNAMIC" == "ON" ]; then $PYTHON2 -c 'import neuron; neuron.test()'; $PYTHON3 -c 'import neuron; neuron.test()'; else if [ "$CMAKE_OPTION" != "-DNRN_ENABLE_PYTHON=OFF" ]; then $PYTHON --version; $PYTHON -c 'import neuron; neuron.test()'; fi; fi; if [ "$BUILD_MODE" != "cmake" ] ; then $PYTHON share/lib/python/neuron/rxdtests/run_all.py; fi;
2167neurondemo: command not found
2168Traceback (most recent call last):
2169  File "<string>", line 1, in <module>
2170ImportError: No module named neuron
2171Traceback (most recent call last):
2172  File "<string>", line 1, in <module>
2173ModuleNotFoundError: No module named 'neuron'
2174The command "neurondemo -nogui -c 'demo(4)' -c 'run()' -c 'quit()'; if [ "$NRN_ENABLE_PYTHON_DYNAMIC" == "ON" ]; then $PYTHON2 -c 'import neuron; neuron.test()'; $PYTHON3 -c 'import neuron; neuron.test()'; else if [ "$CMAKE_OPTION" != "-DNRN_ENABLE_PYTHON=OFF" ]; then $PYTHON --version; $PYTHON -c 'import neuron; neuron.test()'; fi; fi; if [ "$BUILD_MODE" != "cmake" ] ; then $PYTHON share/lib/python/neuron/rxdtests/run_all.py; fi;" exited with 0.
2175
2176
2177Done. Your build exited with 0.
```

Even things failed, it returned `0`!

pramodk commented:
I think this should explain what's going on:

```
→ ls asas; echo "A"
ls: asas: No such file or directory
A
→ echo $?
0

→ echo "A"; ls asasa
A
ls: asasa: No such file or directory
→ echo $?
1

→ ls asas; echo "A"
ls: asas: No such file or directory
A
→ echo $?
0

→ ls asas && echo "A"
ls: asas: No such file or directory

→ echo $?
1
``` 

i.e. with `;` operator, the exit code of last command is returned as final exit code.

I am fixing this.

I just realised that there is no OSX build for NEURON. Can we add that similar to coreneuron / nmodl projects?
